### ISSUE TITLE
  enum variants aren't considered to be const during const-evaluation

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#23898]
Spawned off of #5873

``` rust
use self::Category::*;

enum Category {
    A,
    B,
    C,
    NumCategories,
}
const NUM_CATEGORIES: usize = NumCategories as usize;

fn main() {
    let the_categories: [Category; NUM_CATEGORIES] = [A, B, C];
    println!("{:?}", the_categories);
}
```

[playpen](http://is.gd/wHhpXH)

```
<anon>:9:31: 9:44 error: array length constant evaluation error: non-constant path in constant expr [E0250]
<anon>:9 const NUM_CATEGORIES: usize = NumCategories as usize;
                                       ^~~~~~~~~~~~~
<anon>:12:25: 12:51 note: for array length here
<anon>:12     let the_categories: [Category; NUM_CATEGORIES] = [A, B, C];
                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
```

**Comment by pnkfelix** [comment]
(copied labels from #5873.)

**Comment by pnkfelix** [comment]
cc @doy @catamorphism @huonw @tikue @jdm @FlaPer87@nikomatsakis @jaredly @frewsxcv @tommit

(all of the above were commenters on the original ticket, and thus I assume they may be interested in following its progress here.)

**Comment by pmarcelll** [comment]
This also seems to affect const fn.

**Comment by iirelu** [comment]
Would like to throw in my pennies and say that this makes doing a 1:1 port of C code into Rust (an otherwise relatively easy process) slightly annoying when the C code in question uses this trick. I have to calculate the length of the enum, define that as a const of the same name, and update it every time to the enum changes.

**Comment by shepmaster** [comment]
For what it's worth, this does "work" in Rust 1.14 if you give values to the variants:

```rust
use self::Category::*;

#[derive(Debug)]
enum Category {
    A = 0,
    B = 1,
    C = 2,
    NumCategories = 3,
}
const NUM_CATEGORIES: usize = NumCategories as usize;

fn main() {
    let the_categories: [Category; NUM_CATEGORIES] = [A, B, C];
    println!("{:?}", the_categories);
}
```

However, it continues to fail if you cross the crate boundary (an example is available in #38875)

**Comment by shepmaster** [comment]
It does seem odd and less-than-intuitive that implicit values and explicit values are treated differently, and that intra-crate and inter-crate are treated differently.
